Multisig/UI/Safe Management/Create Safe/Web3AuthMFAService.swift

```diff
@@ -0,0 +1,212 @@
+    private let keychainService: SecureStore!
+    private let postBoxKey: String
+    private let publicAddress: String
+    private var thresholdKey: ThresholdKey
+    private var threshold = 0
+    private let question = "PASSWORD"
+
+    static func mfaEnabled(userInfo: [String: Any]) -> Bool {
+        // TODO: in future version of the SDK there will be an entry called "nonce". If nonce == 0, then mfa is enabled
+
+        // For now we whitelist adresses
+        let email = userInfo["email"] as? String
+
+        return email == "ego.other@gmail.com" || email == "dirk@safe.global"
+    }
+
+    func setupPassword(password: String) async throws {
+        let _ = try await SecurityQuestionModule.generate_new_share(threshold_key: thresholdKey, questions: question, answer: password)
+    }
+
+    /// this is used to unlock the key. Can also be done in init()
+    func inputPassword(password: String) async throws {
+        self.password = password
+
+        guard let result = try? await SecurityQuestionModule.input_share(threshold_key: thresholdKey, answer: password) else {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "input share failed. Make sure threshold key is initialized")
+        }
+        if !result {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "password incorrect")
+        }
+    }
+
+    func changePassword(oldPassword: String, newPassword: String) async throws {
+        // password given
+        guard let result = try? await SecurityQuestionModule.input_share(threshold_key: thresholdKey, answer: oldPassword) else {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "input share failed. Make sure threshold key is initialized")
+        }
+        if !result {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "old password incorrect")
+        }
+
+        guard let reconstructionDetails = try? await thresholdKey.reconstruct() else {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "Failed to reconstruct key. More share(s) required.")
+        }
+
+        let changeResult = try await SecurityQuestionModule.change_question_and_answer(threshold_key: thresholdKey, questions: question, answer: newPassword)
+        if !changeResult {
+            throw GSError.Web3AuthKeyReconstructionError(underlyingError: "Password not changed")
+        }
+    }
+
+    /// This is used on a new device after the user logged in with their social login and we detect, that we cannot
+    /// recreate the final key because we're missing the device share. We need to ask the user for their password
+    /// to recreate the device share and reconstruct the final key
+    func recreateDeviceShare(password: String? = nil) async throws {
+        if let password = password {
+            guard let result = try? await SecurityQuestionModule.input_share(threshold_key: thresholdKey, answer: password) else {
```

Comment:
I noticed that this and other lines have an optional `try?` which skips the error thrown from the `SecurityQuestionModule`. Would it be useful to get that error instead and use it as an underlying error?

Reply:
I will go through the errors. The one that can be thrown from input_share is not helpful: `RuntimeError("Error in SecurityQuestionModule, input_share")`

Reply:
I am not sure the underlying errors are more helpful. Usually the are even more generic than this.

Reply:
the point was that if we skip the errors here then when debugging we won't know what to look for.